File: home_client.go
package main

import (
	"bufio"
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"flag"
	"log"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	MsgRegister  = 0x01
	MsgKeepalive = 0x02
	MsgAck       = 0x03
	MsgDataFwd   = 0x10
	MsgDataReply = 0x11

	KeepaliveInt = 20 * time.Second
	AckTimeout   = 90 * time.Second
	PeerTTL      = 120 * time.Second
)

var Magic = []byte{0xDE, 0xAD, 0xBE, 0xEF}

type PseudoPeer struct {
	ID       uint32
	ExtAddr  string
	Conn     *net.UDPConn
	LastSeen time.Time
}

type Client struct {
	mu         sync.RWMutex
	vpsAddr    *net.UDPAddr
	wgLocal    *net.UDPAddr
	authKey    []byte
	tunnelConn *net.UDPConn
	registered bool
	lastAck    time.Time
	peers      map[uint32]*PseudoPeer
}

func NewClient(vpsStr string, wgLocalStr string, authKey string) *Client {
	vpsAddr, err := net.ResolveUDPAddr("udp", vpsStr)
	if err != nil {
		log.Fatalf("Failed to resolve VPS address: %v", err)
	}
	wgAddr, err := net.ResolveUDPAddr("udp", wgLocalStr)
	if err != nil {
		log.Fatalf("Failed to resolve WG Local address: %v", err)
	}

	return &Client{
		vpsAddr: vpsAddr,
		wgLocal: wgAddr,
		authKey: []byte(authKey),
		peers:   make(map[uint32]*PseudoPeer),
	}
}

func signPacket(data []byte, key []byte) []byte {
	ts := uint64(time.Now().UnixMilli())
	var tsBytes [8]byte
	binary.BigEndian.PutUint64(tsBytes[:], ts)

	msg := make([]byte, len(data)+8)
	copy(msg, data)
	copy(msg[len(data):], tsBytes[:])

	mac := hmac.New(sha256.New, key)
	mac.Write(msg)
	signature := mac.Sum(nil)

	return append(msg, signature...)
}

func verifyPacket(data []byte, key []byte) ([]byte, bool) {
	if len(data) < 40 {
		return nil, false
	}
	macOffset := len(data) - 32
	message := data[:macOffset]
	signature := data[macOffset:]

	mac := hmac.New(sha256.New, key)
	mac.Write(message)
	expectedMAC := mac.Sum(nil)

	if !hmac.Equal(signature, expectedMAC) {
		return nil, false
	}

	tsOffset := macOffset - 8
	ts := binary.BigEndian.Uint64(message[tsOffset:macOffset])
	now := uint64(time.Now().UnixMilli())

	diff := int64(now) - int64(ts)
	if diff < -5000 || diff > 10000 {
		return nil, false
	}

	return message[:tsOffset], true
}

func (c *Client) run(localPort int) {
	conn, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.ParseIP("0.0.0.0"), Port: localPort})
	if err != nil {
		log.Fatalf("Failed to listen on tunnel port: %v", err)
	}
	c.tunnelConn = conn

	c.tunnelConn.SetReadBuffer(4 * 1024 * 1024)
	c.tunnelConn.SetWriteBuffer(4 * 1024 * 1024)

	log.Printf("Home client VPS=%s WG=%s", c.vpsAddr, c.wgLocal)
	log.Printf("Tunnel bound to %s", c.tunnelConn.LocalAddr())

	go c.keepaliveLoop()
	go c.cleanupLoop()

	buf := make([]byte, 65536)
	for {
		n, _, err := c.tunnelConn.ReadFromUDP(buf)
		if err != nil {
			log.Printf("Tunnel read err: %v", err)
			continue
		}
		if n < 4 {
			continue
		}

		if bytes.Equal(buf[:4], Magic) {
			if n < 5 {
				continue
			}
			msgType := buf[4]
			
			var payload []byte
			var ok bool
			var wgBytes []byte

			if msgType == MsgDataFwd {
				if n < 55 { // MAGIC(4) + TYPE(1) + PEER(10) + TS(8) + HMAC(32)
					continue
				}
				payload, ok = verifyPacket(buf[:55], c.authKey)
				wgBytes = buf[55:n]
			} else {
				payload, ok = verifyPacket(buf[:n], c.authKey)
			}

			if !ok {
				continue
			}

			if msgType == MsgAck {
				c.mu.Lock()
				if !c.registered {
					log.Printf("VPS ACK — registered ✓")
					c.registered = true
				}
				c.lastAck = time.Now()
				c.mu.Unlock()
			} else if msgType == MsgDataFwd {
				peerID := binary.BigEndian.Uint32(payload[5:9])
				ip := net.IP(payload[9:13])
				port := binary.BigEndian.Uint16(payload[13:15])
				extAddr := (&net.UDPAddr{IP: ip, Port: int(port)}).String()

				c.mu.Lock()
				peer, exists := c.peers[peerID]
				if !exists {
					pConn, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0})
					if err == nil {
						pConn.SetReadBuffer(4 * 1024 * 1024)
						peer = &PseudoPeer{
							ID:      peerID,
							ExtAddr: extAddr,
							Conn:    pConn,
						}
						c.peers[peerID] = peer
						log.Printf("New peer id=%d ext=%s pseudo=%s", peerID, extAddr, pConn.LocalAddr())
						go c.pseudoPeerReadLoop(peer)
					}
				}
				if peer != nil {
					peer.LastSeen = time.Now()
				}
				c.mu.Unlock()

				if peer != nil {
					peer.Conn.WriteToUDP(wgBytes, c.wgLocal)
				}
			}
		}
	}
}

func (c *Client) pseudoPeerReadLoop(peer *PseudoPeer) {
	buf := make([]byte, 65536)

	// Prebuild reply header: MAGIC(4) | MsgDataReply(1) | ID(4)
	hdr := make([]byte, 9)
	copy(hdr[0:4], Magic)
	hdr[4] = MsgDataReply
	binary.BigEndian.PutUint32(hdr[5:9], peer.ID)

	for {
		n, _, err := peer.Conn.ReadFromUDP(buf)
		if err != nil {
			return
		}

		c.mu.Lock()
		peer.LastSeen = time.Now()
		c.mu.Unlock()

		signedHdr := signPacket(hdr, c.authKey)
		reply := make([]byte, len(signedHdr)+n)
		copy(reply, signedHdr)
		copy(reply[len(signedHdr):], buf[:n])

		c.tunnelConn.WriteToUDP(reply, c.vpsAddr)
	}
}

func (c *Client) keepaliveLoop() {
	regPkt := []byte{0xDE, 0xAD, 0xBE, 0xEF, MsgRegister}
	keepPkt := []byte{0xDE, 0xAD, 0xBE, 0xEF, MsgKeepalive}

	for {
		c.mu.Lock()
		reg := c.registered
		ackElapsed := time.Since(c.lastAck)
		c.mu.Unlock()

		if !reg {
			signedPkt := signPacket(regPkt, c.authKey)
			c.tunnelConn.WriteToUDP(signedPkt, c.vpsAddr)
			log.Printf("REGISTER -> %s", c.vpsAddr)
		} else {
			signedPkt := signPacket(keepPkt, c.authKey)
			c.tunnelConn.WriteToUDP(signedPkt, c.vpsAddr)
			if ackElapsed > AckTimeout {
				log.Printf("No ACK for %v — re-registering...", AckTimeout)
				c.mu.Lock()
				c.registered = false
				c.mu.Unlock()
			}
		}
		time.Sleep(KeepaliveInt)
	}
}

func (c *Client) cleanupLoop() {
	ticker := time.NewTicker(30 * time.Second)
	for range ticker.C {
		c.mu.Lock()
		now := time.Now()
		for id, p := range c.peers {
			if now.Sub(p.LastSeen) > PeerTTL {
				p.Conn.Close()
				delete(c.peers, id)
				log.Printf("Closed idle pseudo-peer id=%d", id)
			}
		}
		c.mu.Unlock()
	}
}

func parseINI(filename string, section string) map[string]string {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatalf("Failed to open config %s: %v", filename, err)
	}
	defer file.Close()

	config := make(map[string]string)
	scanner := bufio.NewScanner(file)
	currentSection := ""

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if len(line) == 0 || strings.HasPrefix(line, "#") || strings.HasPrefix(line, ";") {
			continue
		}
		if strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") {
			currentSection = line[1 : len(line)-1]
			continue
		}
		if currentSection == section {
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				key := strings.TrimSpace(parts[0])
				valPart := parts[1]
				if idx := strings.IndexAny(valPart, "#;"); idx != -1 {
					valPart = valPart[:idx]
				}
				config[key] = strings.TrimSpace(valPart)
			}
		}
	}
	return config
}

func main() {
	cfgPath := flag.String("config", "config.ini", "Path to config file")
	flag.Parse()

	log.SetFlags(log.LstdFlags | log.Lmicroseconds)

	config := parseINI(*cfgPath, "client")

	vpsIP, ok := config["vps_ip"]
	if !ok || vpsIP == "" {
		log.Fatalf("Missing 'vps_ip' in [client] section of %s", *cfgPath)
	}

	wgPort := "51820"
	if p, ok := config["wg_port"]; ok && p != "" {
		wgPort = p
	}
	vpsEndpoint := vpsIP + ":" + wgPort

	wgLocalPort := "51820"
	if p, ok := config["wg_local_port"]; ok && p != "" {
		wgLocalPort = p
	}
	wgLocal := "127.0.0.1:" + wgLocalPort

	localPort := 0
	if pStr, ok := config["local_bind_port"]; ok {
		if p, err := strconv.Atoi(pStr); err == nil {
			localPort = p
		}
	}
	
	authKey, ok := config["auth_key"]
	if !ok || authKey == "" || authKey == "CHANGE_ME_TO_A_SECURE_RANDOM_STRING" {
		log.Fatalf("Missing or default 'auth_key' in [client] section of %s", *cfgPath)
	}

	client := NewClient(vpsEndpoint, wgLocal, authKey)
	client.run(localPort)
}

File: home_client.py
#!/usr/bin/env python3
"""
home_client.py — Hole-punch client (runs on home server behind CGNAT)
======================================================================
1. Opens a UDP socket → connects to VPS:ctrl_port
   (this outbound packet creates the NAT mapping in your CGNAT)
2. Sends REGISTER so the VPS records your NAT endpoint
3. Sends KEEPALIVE every 20 s to prevent NAT mapping expiry
4. When VPS forwards a WireGuard packet:
     → creates a pseudo-socket for that external peer
     → delivers raw WG bytes to local WireGuard at 127.0.0.1:wg_local_port
     → when WireGuard replies, wraps and sends back to VPS
5. Logs everything at DEBUG level so you can watch traffic in real time

Run:
  python home_client.py [config.ini]

Press Ctrl+C to stop.
"""

import asyncio
import logging
import socket
import struct
import sys
import time
from dataclasses import dataclass, field
from typing import Dict, Optional, Tuple

# ── Logging ──────────────────────────────────────────────────────────────────

logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s  %(levelname)-8s  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("home")

# ── Wire protocol (must match vps_relay.py) ───────────────────────────────────

MAGIC          = b"\xDE\xAD\xBE\xEF"
MSG_REGISTER   = 0x01
MSG_KEEPALIVE  = 0x02
MSG_ACK        = 0x03
MSG_DATA_FWD   = 0x10
MSG_DATA_REPLY = 0x11

KEEPALIVE_INTERVAL = 20   # s  — must be less than your CGNAT's UDP idle timeout
ACK_TIMEOUT        = 90   # s  — if VPS goes quiet, re-register
PEER_TTL           = 120  # s  — close idle pseudo-sockets after this


# ── Per-external-peer pseudo-socket ───────────────────────────────────────────

class _PseudoProto(asyncio.DatagramProtocol):
    """
    A tiny local UDP socket sitting between the home_client and WireGuard.
    WireGuard replies here; we forward them back to the VPS tunnel.
    """
    def __init__(self, peer_id: int, on_reply):
        self.peer_id  = peer_id
        self._reply   = on_reply   # fn(peer_id, bytes)

    def connection_made(self, transport):
        self.transport = transport

    def datagram_received(self, data: bytes, addr):
        log.debug(f"WG→relay  id={self.peer_id}  {len(data)} B")
        self._reply(self.peer_id, data)

    def error_received(self, exc):
        log.error(f"pseudo-peer {self.peer_id} error: {exc}")


@dataclass
class PseudoPeer:
    peer_id:   int
    peer_addr: Tuple[str, int]   # original external peer (logging only)
    transport: object = field(default=None, repr=False)
    last_seen: float  = field(default_factory=time.monotonic)


# ── Main client ───────────────────────────────────────────────────────────────

class HomeClient:
    def __init__(self, cfg: dict):
        self.vps_addr   = (cfg["vps_ip"], cfg["ctrl_port"])
        self.wg_local   = ("127.0.0.1",  cfg["wg_local_port"])
        self.local_port = cfg["local_bind_port"]

        self.registered  = False
        self.last_ack    = 0.0
        self.tunnel_tx   = None

        self._peers: Dict[int, PseudoPeer] = {}
        self._loop: Optional[asyncio.AbstractEventLoop] = None

    # ── Outbound packets ──────────────────────────────────────────────────────

    def _send(self, msg_type: int, payload: bytes = b""):
        pkt = MAGIC + bytes([msg_type]) + payload
        self.tunnel_tx.sendto(pkt, self.vps_addr)

    def _register(self):
        self._send(MSG_REGISTER)
        log.info(f"REGISTER → {self.vps_addr}")

    def _keepalive(self):
        self._send(MSG_KEEPALIVE)
        log.debug("KEEPALIVE →")

    def _reply_to_vps(self, peer_id: int, wg_bytes: bytes):
        hdr = struct.pack("!I", peer_id)
        self._send(MSG_DATA_REPLY, hdr + wg_bytes)
        log.debug(f"relay→VPS  id={peer_id}  {len(wg_bytes)} B")

    # ── Tunnel protocol (this socket ↔ VPS) ──────────────────────────────────

    class _Tunnel(asyncio.DatagramProtocol):
        def __init__(self, c): self.c = c
        def connection_made(self, tx):
            self.c.tunnel_tx = tx
            sock = tx.get_extra_info("sockname")
            log.info(f"Tunnel socket  local={sock[0]}:{sock[1]}  vps={self.c.vps_addr}")
            self.c._register()
        def datagram_received(self, data, addr):
            self.c._on_vps(data)
        def error_received(self, exc):
            log.error(f"tunnel error: {exc}")
        def connection_lost(self, exc):
            log.warning(f"tunnel lost: {exc}")
            self.c.registered = False

    def _on_vps(self, data: bytes):
        if len(data) < 5 or data[:4] != MAGIC:
            return
        t       = data[4]
        payload = data[5:]

        if t == MSG_ACK:
            if not self.registered:
                log.info("VPS ACK received — registered ✓")
                self.registered = True
            self.last_ack = time.monotonic()

        elif t == MSG_DATA_FWD:
            # payload: peer_id(4) | src_ip(4) | src_port(2) | wg_bytes
            if len(payload) < 11:
                return
            peer_id   = struct.unpack("!I", payload[:4])[0]
            peer_ip   = socket.inet_ntoa(payload[4:8])
            peer_port = struct.unpack("!H", payload[8:10])[0]
            wg_bytes  = payload[10:]
            asyncio.ensure_future(
                self._to_wg(peer_id, (peer_ip, peer_port), wg_bytes)
            )

    # ── Relay to local WireGuard ──────────────────────────────────────────────

    async def _to_wg(self, peer_id: int, peer_addr: Tuple[str, int], wg_bytes: bytes):
        if peer_id not in self._peers:
            proto  = _PseudoProto(peer_id, self._reply_to_vps)
            tx, _  = await self._loop.create_datagram_endpoint(
                lambda: proto,
                local_addr  = ("127.0.0.1", 0),
                remote_addr = self.wg_local,
            )
            local_port = tx.get_extra_info("sockname")[1]
            peer = PseudoPeer(peer_id=peer_id, peer_addr=peer_addr, transport=tx)
            self._peers[peer_id] = peer
            log.info(
                f"New peer  id={peer_id}  ext={peer_addr}  "
                f"local_pseudo=127.0.0.1:{local_port}"
            )
        else:
            peer = self._peers[peer_id]
            peer.last_seen = time.monotonic()

        peer.transport.sendto(wg_bytes)
        log.debug(f"VPS→WG  id={peer_id}  {len(wg_bytes)} B")

    # ── Background tasks ──────────────────────────────────────────────────────

    async def _keepalive_loop(self):
        # Retry registration for up to 60 s
        for _ in range(12):
            await asyncio.sleep(5)
            if self.registered:
                break
            log.warning("Not registered yet — retrying…")
            self._register()
        if not self.registered:
            log.error(
                "Could not reach VPS after 60 s.\n"
                "  • Check vps_ip / ctrl_port in config.ini\n"
                "  • Check AWS Security Group: UDP 51821 open inbound\n"
                "  • Is vps_relay.py running on the VPS?"
            )

        while True:
            await asyncio.sleep(KEEPALIVE_INTERVAL)
            if not self.registered:
                log.warning("Lost registration — re-registering…")
                self._register()
            else:
                self._keepalive()
                if time.monotonic() - self.last_ack > ACK_TIMEOUT:
                    log.warning(f"No ACK for {ACK_TIMEOUT} s — re-registering…")
                    self.registered = False
                    self._register()

    async def _cleanup_loop(self):
        while True:
            await asyncio.sleep(60)
            now   = time.monotonic()
            stale = [p for p in list(self._peers.values())
                     if now - p.last_seen > PEER_TTL]
            for p in stale:
                p.transport.close()
                del self._peers[p.peer_id]
                log.info(f"Closed idle pseudo-peer  id={p.peer_id}  ext={p.peer_addr}")

    # ── Run ───────────────────────────────────────────────────────────────────

    async def run(self):
        self._loop = asyncio.get_running_loop()
        log.info(f"Home client  VPS={self.vps_addr}  WG={self.wg_local}")

        transport, _ = await self._loop.create_datagram_endpoint(
            lambda: self._Tunnel(self),
            local_addr  = ("0.0.0.0", self.local_port),
            remote_addr = self.vps_addr,
        )

        try:
            await asyncio.gather(
                self._keepalive_loop(),
                self._cleanup_loop(),
            )
        finally:
            transport.close()
            for p in self._peers.values():
                p.transport.close()


# ── Config ────────────────────────────────────────────────────────────────────

def load_config(path: str = "config.ini") -> dict:
    import configparser
    cfg = configparser.ConfigParser()
    cfg.read(path)
    if "client" not in cfg:
        sys.exit(f"[ERROR] Missing [client] section in {path}")
    s = cfg["client"]
    if not s.get("vps_ip", "").strip():
        sys.exit("[ERROR] 'vps_ip' must be set in [client] section")
    return {
        "vps_ip":          s["vps_ip"].strip(),
        "ctrl_port":       int(s.get("ctrl_port",       "51821")),
        "wg_local_port":   int(s.get("wg_local_port",  "51820")),
        "local_bind_port": int(s.get("local_bind_port", "0")),
    }


# ── Main ──────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    cfg_path = sys.argv[1] if len(sys.argv) > 1 else "config.ini"
    try:
        asyncio.run(HomeClient(load_config(cfg_path)).run())
    except KeyboardInterrupt:
        log.info("Stopped.")
